File: src/com/l2kt/gameserver/geoengine/geodata/GeoLocation.java
package com.l2kt.gameserver.geoengine.geodata;

import com.l2kt.gameserver.geoengine.GeoEngine;
import com.l2kt.gameserver.model.location.Location;

public class GeoLocation extends Location
{
	private byte _nswe;
	
	public GeoLocation(int x, int y, int z)
	{
		super(x, y, GeoEngine.getInstance().getHeightNearest(x, y, z));
		_nswe = GeoEngine.getInstance().getNsweNearest(x, y, z);
	}
	
	public void set(int x, int y, short z)
	{
		super.set(x, y, GeoEngine.getInstance().getHeightNearest(x, y, z));
		_nswe = GeoEngine.getInstance().getNsweNearest(x, y, z);
	}
	
	public int getGeoX()
	{
		return _x;
	}
	
	public int getGeoY()
	{
		return _y;
	}
	
	@Override
	public int getX()
	{
		return GeoEngine.getWorldX(_x);
	}
	
	@Override
	public int getY()
	{
		return GeoEngine.getWorldY(_y);
	}
	
	public byte getNSWE()
	{
		return _nswe;
	}
}

File: src/com/l2kt/gameserver/geoengine/geodata/GeoLocation.kt
package com.l2kt.gameserver.geoengine.geodata

import com.l2kt.gameserver.geoengine.GeoEngine
import com.l2kt.gameserver.model.location.Location

class GeoLocation(x: Int, y: Int, z: Int) : Location(x, y, GeoEngine.getHeightNearest(x, y, z).toInt()) {
    var nswe: Byte = GeoEngine.getNsweNearest(x, y, z)

    val geoX: Int
        get() = super.x

    val geoY: Int
        get() = super.y

    override var x: Int
        get() = GeoEngine.getWorldX(super.x)
        set(value) {
            super.x = value
        }

    override var y: Int
        get() = GeoEngine.getWorldY(super.y)
        set(value) {
            super.y = value
        }

    operator fun set(x: Int, y: Int, z: Short) {
        super.set(x, y, GeoEngine.getHeightNearest(x, y, z.toInt()).toInt())
        nswe = GeoEngine.getNsweNearest(x, y, z.toInt())
    }
}